client/pages/chat-page.jsx

```diff
@@ -1,23 +1,90 @@
 import React from 'react';
 import { io } from 'socket.io-client';
+import AppContext from '../lib/app-context';
+import ChatMessage from '../components/chat-message';
+import { format } from 'date-fns';
 
 export default class ChatMain extends React.Component {
   constructor(props) {
     super(props);
     this.state = {
-      chatMsg: null
+      chatMsg: [],
+      users: [],
+      input: '',
+      typingMsg: ''
     };
+    this.handleChange = this.handleChange.bind(this);
+    this.handleSubmit = this.handleSubmit.bind(this);
+    this.handleType = this.handleType.bind(this);
+    this.textBox = React.createRef();
   }
 
   componentDidMount() {
+    const username = this.context.user.username;
     this.socket = io({ query: `roomName=${this.props.roomName}` });
+    this.socket.emit('joinRoom', { username });
+    this.socket.on('roomUsers', data => {
+      this.setState({ users: data });
+    });
+    this.socket.on('message', message => {
+      this.setState({ chatMsg: this.state.chatMsg.concat(message), typingMsg: '' });
+    });
+
+    this.socket.on('typing', message => {
+      this.setState({ typingMsg: `${message.username}${message.text}` });
+    });
+
+  }
+
+  handleChange(event) {
+    this.setState({ input: event.target.value });
+  }
+
+  handleSubmit(event) {
+    event.preventDefault();
+    if (this.state.input) {
+      const newMsg = this.state.input.trim();
+
+      this.socket.emit('messageChat', newMsg);
+      this.setState({ input: '' });
+    }
+
+  }
+
+  handleType(event) {
+    this.socket.emit('typing');
+  }
+
+  componentDidUpdate() {
+    this.textBox.current.scrollTop = this.textBox.current.scrollHeight;
+
```

Comment:
It's risky to do `componentDidUpdate` with no condition. I recommend checking the messages state and comparing to the previous messages state.